File: js/plugins/middleware/src/fallback.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  GENKIT_UI_METADATA,
  GENKIT_UI_WIDGETS,
  GenkitError,
  ModelReferenceSchema,
  StatusNameSchema,
  annotateSchema,
  generateMiddleware,
  z,
  type GenerateMiddleware,
  type StatusName,
} from 'genkit';
import { logger } from 'genkit/logging';
import { ModelAction } from 'genkit/model';
import { Registry } from 'genkit/registry';

const DEFAULT_FALLBACK_STATUSES: StatusName[] = [
  'UNAVAILABLE',
  'DEADLINE_EXCEEDED',
  'RESOURCE_EXHAUSTED',
  'ABORTED',
  'INTERNAL',
  'NOT_FOUND',
  'UNIMPLEMENTED',
];

export const FallbackOptionsSchema = z
  .object({
    /**
     * An array of models to try in order.
     */
    models: annotateSchema(
      z
        .array(ModelReferenceSchema)
        .describe('An array of models to try in order.'),
      { [GENKIT_UI_METADATA.WIDGET]: GENKIT_UI_WIDGETS.MODEL_LIST }
    ),
    /**
     * An array of `StatusName` values that should trigger a fallback.
     * @default ['UNAVAILABLE', 'DEADLINE_EXCEEDED', 'RESOURCE_EXHAUSTED', 'ABORTED', 'INTERNAL', 'NOT_FOUND', 'UNIMPLEMENTED']
     */
    statuses: z
      .array(StatusNameSchema)
      .optional()
      .describe(
        'An array of StatusName values that should trigger a fallback.'
      ),
    /**
     * If true, the fallback model will not inherit the original request's configuration.
     * @default false
     */
    isolateConfig: z
      .boolean()
      .optional()
      .describe(
        "If true, the fallback model will not inherit the original request's configuration."
      ),
  })
  .passthrough();

export type FallbackOptions = z.infer<typeof FallbackOptionsSchema>;

/**
 * Creates a middleware that falls back to a different model on specific error statuses.
 *
 * ```ts
 * const { text } = await ai.generate({
 *   model: googleAI.model('gemini-2.5-pro'),
 *   prompt: 'You are a helpful AI assistant named Walt, say hello',
 *   use: [
 *     fallback({
 *       models: [googleAI.model('gemini-2.5-flash')],
 *       statuses: ['RESOURCE_EXHAUSTED'],
 *     }),
 *   ],
 * });
 * ```
 */
export const fallback: GenerateMiddleware<typeof FallbackOptionsSchema> =
  generateMiddleware(
    {
      name: 'fallback',
      description: 'Fallback to a different model on specific error statuses.',
      configSchema: FallbackOptionsSchema,
    },
    (options) => {
      const {
        models = [],
        statuses = DEFAULT_FALLBACK_STATUSES,
        isolateConfig = false,
      } = options.config || {};

      return {
        model: async (req, ctx, next) => {
          try {
            return await next(req, ctx);
          } catch (e: any) {
            const isFallbackable =
              e instanceof GenkitError &&
              statuses.includes(e.status as StatusName);
            if (isFallbackable) {
              let lastError: any = e;
              for (const model of models) {
                logger.logStructuredWarn(
                  `Request failed with status ${lastError.status}: ${lastError.message}. Falling back to model ${model.name}...`,
                  {
                    'genkit.middleware.name': 'fallback',
                    'genkit.middleware.fallback.target_model': model.name,
                  },
                  lastError
                );
                const normalizedModel = await resolveModel(
                  options.ai.registry,
                  model
                );
                try {
                  return await normalizedModel.model(
                    {
                      ...req,
                      config: isolateConfig
                        ? normalizedModel.config
                        : (normalizedModel.config ?? req.config),
                    },
                    ctx
                  );
                } catch (e2: any) {
                  lastError = e2;
                  if (
                    e2 instanceof GenkitError &&
                    statuses.includes(e2.status as StatusName)
                  ) {
                    continue;
                  }
                  const statusStr =
                    e2 instanceof GenkitError
                      ? `status ${e2.status}`
                      : 'non-Genkit error';
                  logger.logStructuredWarn(
                    `Fallback model ${model.name} failed with ${statusStr}: ${e2.message || String(e2)}. Aborting fallback sequence.`,
                    {
                      'genkit.middleware.name': 'fallback',
                      'genkit.middleware.fallback.target_model': model.name,
                    },
                    e2
                  );
                  throw e2;
                }
              }
              logger.logStructuredWarn(
                `All fallback options exhausted. Last error: ${lastError.message || String(lastError)}`,
                {
                  'genkit.middleware.name': 'fallback',
                },
                lastError
              );
              throw lastError;
            } else {
              const statusStr =
                e instanceof GenkitError
                  ? `status ${e.status}`
                  : 'non-Genkit error';
              logger.logStructuredWarn(
                `Request failed with ${statusStr}: ${e.message || String(e)}. Skipping fallback.`,
                {
                  'genkit.middleware.name': 'fallback',
                },
                e
              );
              throw e;
            }
          }
        },
      };
    }
  );

async function resolveModel(
  registry: Registry,
  model: z.infer<typeof ModelReferenceSchema>
): Promise<{ model: ModelAction; config?: any }> {
  return {
    model: await registry.lookupAction(`/model/${model.name}`),
    config: model.config,
  };
}

File: go/plugins/middleware/fallback.go
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package middleware

import (
	"context"
	"slices"

	genkit "github.com/firebase/genkit/go"
	"github.com/firebase/genkit/go/ai"
	"github.com/firebase/genkit/go/core/status"
)

// defaultFallbackStatuses are the status codes that trigger a fallback by default.
var defaultFallbackStatuses = []status.Name{
	status.Unavailable,
	status.DeadlineExceeded,
	status.ResourceExhausted,
	status.Aborted,
	status.Internal,
	status.NotFound,
	status.Unimplemented,
}

// Fallback is a middleware that tries alternative models when the primary model
// fails with a retryable error status.
//
// It only hooks the Model stage -- when a model API call fails with a matching
// status, the request is forwarded to the next model in the list.
//
// Models are specified as [ai.ActionRef] values (created via [ai.NewActionRef])
// and resolved via the [genkit.Genkit] instance at call time.
//
// Usage:
//
//	resp, err := g.Generate(ctx,
//	    ai.WithModel(primary),
//	    ai.WithPrompt("hello"),
//	    ai.WithUse(&middleware.Fallback{Models: []ai.ActionRef{
//	        googlegenai.ModelRef("googleai/gemini-2.5-flash", ...),
//	        googlegenai.ModelRef("vertexai/gemini-2.5-flash", ...),
//	    }}),
//	)
type Fallback struct {
	// Models is the ordered list of fallback models to try.
	// These are tried in order after the primary model fails. Each ref's
	// Config is used verbatim for that model -- the original request's
	// Config is not inherited. Use [ai.NewActionRef] to attach config.
	Models []ai.ActionRef `json:"models,omitempty"`
	// Statuses is the set of status codes that trigger a fallback. An error
	// carrying no status counts as INTERNAL for this check.
	// Defaults to [defaultFallbackStatuses].
	Statuses []status.Name `json:"statuses,omitempty"`
}

func (f *Fallback) Name() string { return provider + "/fallback" }

func (f *Fallback) New(ctx context.Context) (*ai.Hooks, error) {
	return &ai.Hooks{
		WrapModel: f.wrapModel,
	}, nil
}

func (f *Fallback) statuses() []status.Name {
	if len(f.Statuses) > 0 {
		return f.Statuses
	}
	return defaultFallbackStatuses
}

func (f *Fallback) wrapModel(ctx context.Context, params *ai.ModelParams, next ai.ModelNext) (*ai.ModelResponse, error) {
	resp, err := next(ctx, params)
	if err == nil {
		return resp, nil
	}

	if !isFallbackRetryable(err, f.statuses()) {
		return nil, err
	}

	lastErr := err
	for _, ref := range f.Models {
		name := ref.Name()
		m := genkit.FromContext(ctx).LookupModel(name)
		if m == nil {
			return nil, status.Errorf(ai.ErrModelNotFound, "fallback: model %q not found", name)
		}
		req := *params.Request
		req.Config = ref.Config()
		resp, err := m.Generate(ctx, &req, params.Callback)
		if err == nil {
			return resp, nil
		}
		lastErr = err
		if !isFallbackRetryable(err, f.statuses()) {
			return nil, err
		}
	}
	return nil, lastErr
}

// isFallbackRetryable reports whether err should trigger trying the next model:
// its status must be in the provided list. An error carrying no status counts
// as INTERNAL, which the default list includes, so an unclassified provider
// failure still falls through to the next model.
func isFallbackRetryable(err error, statuses []status.Name) bool {
	return slices.Contains(statuses, status.Of(err))
}
